src/Generators/ControllerTagGenerator.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace SilverLeague\IDEAnnotator;
+
+use Psr\Container\NotFoundExceptionInterface;
+use ReflectionClass;
+use ReflectionException;
+use SilverStripe\CMS\Controllers\ContentController;
+use SilverStripe\Core\ClassInfo;
+
+class ControllerTagGenerator extends AbstractTagGenerator
+{
+
+    /**
+     * @return void
+     * @throws NotFoundExceptionInterface
+     */
+    protected function generateTags()
+    {
+        $this->generateControllerObjectTags();
+        $this->generateExtensionsTags();
+        $this->generateOwnerTags();
+    }
+
+    protected function generateControllerObjectTags()
+    {
+        $pageClassname = str_replace(['_Controller', 'Controller'], '', $this->className);
+        if (class_exists($pageClassname) && $this->isContentController($this->className)) {
+            $this->pushPropertyTag("\\$pageClassname" . ' dataRecord');
+            $this->pushMethodTag($pageClassname, "\\$pageClassname" . ' data()');
+
+            // don't mixin Page, since this is a ContentController method
+            if ($pageClassname !== 'Page') {
+                $this->pushMixinTag("\\$pageClassname");
+            }
+        }
+    }
+
+    /**
+     * @param string $className
+     * @return bool
+     * @throws ReflectionException
+     */
+    protected function isContentController($className)
+    {
+        $reflector = new ReflectionClass($className);
+
+        return ClassInfo::exists(ContentController::class)
```

Comment:
We can use `class_exists()` now since SS is vendorising everything, so the Composer autoloader is a first class citizen:

```php
return class_exists(ContentController::class) && singleton($className) instanceof ContentController;
```

Reply:
That API might get deprecated in 4.x but it definitely won't be removed before 5

Reply:
Hmmm, I thought it was the other way around and SilverStripe prefers this way? Low prio fix though, could you file an issue for it?

Reply:
I think it's fine :) someone will update it eventually when it gets deprecated